tools/antithesis-go-instrumentor/antithesis-go-instrumentor.go

```diff
@@ -0,0 +1,227 @@
+
+	cmd := exec.Command("bash", "-c", commandLine)
+	logger.Printf("%s", commandLine)
+	_, err := cmd.Output()
+	if err != nil {
+		// Errors here are pretty mysterious.
+		logger.Fatalf("%v", err)
+	}
+}
+
+func writeInstrumentedSource(source, path string) error {
+	// Any errors here are fatal anyway, so I'm not checking.
+	f, e := os.Create(path)
+	if e != nil {
+		logger.Printf("Error: could not create %s", path)
+		return e
+	}
+	defer f.Close()
+	_, e = f.WriteString(source)
+	if e != nil {
+		logger.Printf("Error: Could not write instrumented source to %s", path)
+		return e
+	}
+	return nil
+}
+
+func main() {
+	cmd_args := parse_args()
+	if cmd_args.ShowVersion {
+		fmt.Println(strings.TrimSpace(versionString))
+		os.Exit(0)
+	}
+
+	if cmd_args.InvalidArgs {
+		os.Exit(1)
+	}
+
+	var err error
+
+	// Setup logging globals
+	// TODO Could also pass these settings inside a parameter (or inside a receiver)
+	logger = cmd_args.Logger
+	verbosity = cmd_args.Verbosity
+
+	logger.Println(strings.TrimSpace(versionString))
+
+	if !IsGoAvailable() {
+		logger.Printf("Go toolchain not available")
+		os.Exit(1)
+	}
+
+	var cmd_files *CommandFiles
+
+	if err, cmd_files = cmd_args.NewCommandFiles(); err != nil {
+		logger.Printf(err.Error())
+		os.Exit(1)
+	}
+
+	if err = cmd_files.ParseExclusionsFile(); err != nil {
+		logger.Printf(err.Error())
+		os.Exit(1)
+	}
+
+	sourceFiles := []string{}
+	if err, sourceFiles = cmd_files.FindSourceCode(); err != nil {
+		logger.Printf(err.Error())
+		os.Exit(1)
+	}
+
+	files_hash := HashFileContent(sourceFiles)[0:12]
+	var instrumentor *Instrumentor = nil
+	var symbolTableWriter *SymbolTable = nil
+
+	if cmd_args.WantsInstrumentor {
+		logger.Printf("Instrumenting %s to %s", cmd_files.InputDirectory, cmd_files.CustomerDirectory)
+		symbolTableWriter = cmd_files.CreateSymbolTableWriter(files_hash)
+		instrumentor = CreateInstrumentor(cmd_files.InputDirectory, InstrumentationModuleName, symbolTableWriter)
+	}
+	usingSymbols := cmd_files.UsingSymbols()
+	full_catalog_path := cmd_files.CatalogPath
+
+	// Setup the assertion scanner (used to create the assertion catalog)
+	aSI := NewScanningInfo(verbosity > 0, full_catalog_path, usingSymbols)
+
+	filesInstrumented := 0
+	filesCataloged := 0
+	previousEdge := 0
+	instrumented := ""
+	var e error = nil
+
+	for _, path := range sourceFiles {
+		base_name := filepath.Base(path)
+		if path_was_generated := strings.HasSuffix(base_name, GENERATED_SUFFIX); path_was_generated {
+			logger.Printf("Skipping %s", path)
+			continue
+		}
+		if cmd_files.wants_instrumentor {
+			logger.Printf("Instrumenting %s", path)
+			previousEdge = instrumentor.CurrentEdge
+			instrumented, e = instrumentor.Instrument(path)
+
+			if e != nil {
+				logger.Printf("Error: File %s produced error %s; simply copying source", path, e)
+				continue
+			}
+
+			if instrumented == "" {
+				// The instrumentor should have reported why it didn't instrument this file.
+				continue
+			}
+		}
```

Comment:
If there's a problem instrumenting, does that mean we don't want to catalog? Couldn't there be code that we can catalog quite fine, but can't instrument?

Reply:
**Fixed previously - yes, assertion scanning happens even when instrumentation fails:

```
	//--------------------------------------------------------------------------------
	// Process all files (ignore previously generated assertion catalogs)
	//--------------------------------------------------------------------------------
	for _, file_name := range source_files {
		if IsGeneratedFile(file_name) {
			logger.Printf("Skipping %s", file_name)
			continue
		}

		if instrumented_source := cI.InstrumentFile(file_name); instrumented_source != "" {
			cmd_files.WriteInstrumentedOutput(file_name, instrumented_source, cI)
		}

		aSI.ScanFile(file_name)
	}
```